src/hotspot/share/prims/jvmtiThreadState.inline.hpp

```diff
@@ -78,18 +78,22 @@ inline JvmtiThreadState* JvmtiThreadState::state_for_while_locked(JavaThread *th
 
   NoSafepointVerifier nsv;  // oop is safe to use.
 
-  if (thread_oop == nullptr) {  // Then thread should not be null (see assert above).
-    thread_oop = thread->jvmti_vthread() != nullptr ? thread->jvmti_vthread() : thread->threadObj();
-  }
-
   // In a case of unmounted virtual thread the thread can be null.
   JvmtiThreadState *state = thread == nullptr ? nullptr : thread->jvmti_thread_state();
 
   if (state == nullptr && thread != nullptr && thread->is_exiting()) {
     // Don't add a JvmtiThreadState to a thread that is exiting.
     return nullptr;
   }
-  if (state == nullptr || state->get_thread_oop() != thread_oop) {
+
+  if (thread_oop == nullptr) {  // Then thread should not be null (see assert above).
+    thread_oop = thread->jvmti_vthread() != nullptr ? thread->jvmti_vthread() : thread->threadObj();
+  }
+  // The state->get_thread_oop() may be null if the state is created during
+  // the allocation of the thread oop when a native thread is attaching. Make
+  // sure we don't create a new state for the JavaThread.
+  if (state == nullptr || (state->get_thread_oop() != nullptr &&
```

Comment:
In my limited testing with and without virtual threads, when `state` is not null, `state->get_thread_oop()` is always either null or equal to `thread_oop`. So I suspect this whole if is equivalent to `if (state == nullptr)` in practice.

I think it is better to split this if into two conditions. Here is my proposed change: https://github.com/openjdk/jdk/commit/00ace66c36243671a0fb1b673b3f9845460c6d22. It adds a `JvmtiThreadState::set_thread_oop()` method, and checks the above observation with `state->get_thread_oop()`, and addresses the problem of `JvmtiThreadState::_thread_oop_h` incorrectly staying null for attached native thread.

I tested it with the 10000 virtual threads running Thread.sleep() example from https://openjdk.org/jeps/425, and  `make test TEST=jdk_loom`.

Reply:
If my understand of [JavaThread::rebind_to_jvmti_thread_state_of](https://github.com/openjdk/jdk/blob/fe0ccdf5f8a5559a608d2e2cd2b6aecbe245c5ec/src/hotspot/share/runtime/javaThread.cpp#L1819) is correct, state->get_thread_oop() could be either `JavaThread::_threadObj` or `JavaThread::_jvmti_vthread`, ignoring the null case here.

The `state` could be null if the virtual thread is unmounted and the thread is null, or if a JvmtiThreadState has not been created for the thread.

Currently the `JvmtiThreadState::_thread_oop_h` is 'sealed' once a JvmtiThreadState is created. It can never be altered before a JvmtiThreadState is destroyed. Adding a `JvmtiThreadState:set_thread_oop()` changes that assumption. Let's wait for others who are more involved in this area to comment as well.

I just took a look of your https://github.com/openjdk/jdk/commit/00ace66c36243671a0fb1b673b3f9845460c6d22 change. 
I think it does not address the specific issue that we were seeing in our tests. At the time when `JvmtiThreadState::state_for_while_locked` is called for a JavaThread associated with a native thread being attached, the thread oop is being allocated. There is no non-null thread oop to be set into `JvmtiThreadState::_thread_oop_h`.

I'll address other comments later. Thanks, @caoman.